File: ts/src/memo/encode.ts
import type { EncodeMemoV1Params } from "../types";
import {
  TYPE_CODES,
  MEMO_BYTES,
  ISSUER_TAG_OFFSET,
  ISSUER_TAG_SIZE,
  ID16_OFFSET,
  SALT_OFFSET,
  SALT_SIZE,
} from "./constants";
import { ulidToBytes16 } from "./ulid";

/**
 * Generate 7 cryptographically random bytes for the salt field.
 *
 * @returns 7-byte Uint8Array from `crypto.getRandomValues`
 * @example
 * ```ts
 * const salt = randomSalt();
 * const memo = encodeMemoV1({ type: "invoice", issuerTag, ulid, salt });
 * ```
 */
export function randomSalt(): Uint8Array {
  const buf = new Uint8Array(SALT_SIZE);
  crypto.getRandomValues(buf);
  return buf;
}

/**
 * Encode memo fields into a 32-byte hex string per TEMPO-RECONCILE-MEMO-001.
 *
 * Layout: `[type:1][issuerTag:8][id16:16][salt:7]` = 32 bytes.
 *
 * @param params - Memo fields: type, issuerTag, ulid, and optional salt
 * @returns `0x`-prefixed 66-character hex string (32 bytes)
 * @throws If `params.type` is not a valid memo type code
 * @throws If `params.salt` is provided and not exactly 7 bytes
 * @example
 * ```ts
 * const memo = encodeMemoV1({
 *   type: "invoice",
 *   issuerTag: issuerTagFromNamespace("my-app"),
 *   ulid: "01MASW9NF6YW40J40H289H858P",
 * });
 * ```
 */
export function encodeMemoV1(params: EncodeMemoV1Params): `0x${string}` {
  const typeCode = TYPE_CODES[params.type];
  if (typeCode === undefined) {
    throw new Error(`Invalid memo type: ${params.type}`);
  }

  const id16 = ulidToBytes16(params.ulid);

  const salt = params.salt === "random" ? randomSalt() : (params.salt ?? new Uint8Array(SALT_SIZE));
  if (salt.length !== SALT_SIZE) {
    throw new Error(`salt must be ${SALT_SIZE} bytes, got ${salt.length}`);
  }

  const buf = new Uint8Array(MEMO_BYTES);

  buf[0] = typeCode;

  const tag = params.issuerTag;
  if (tag < 0n || tag > 0xffffffffffffffffn) {
    throw new Error(`issuerTag must be a uint64 [0, 2^64-1], got ${tag}`);
  }
  for (let i = 0; i < ISSUER_TAG_SIZE; i++) {
    buf[ISSUER_TAG_OFFSET + i] = Number((tag >> BigInt((ISSUER_TAG_SIZE - 1 - i) * 8)) & 0xffn);
  }

  buf.set(id16, ID16_OFFSET);
  buf.set(salt, SALT_OFFSET);

  let hex = "0x";
  for (let i = 0; i < MEMO_BYTES; i++) {
    hex += buf[i]!.toString(16).padStart(2, "0");
  }

  return hex as `0x${string}`;
}

File: rs/tempo-reconcile/src/memo/encode.rs
use super::constants::{
    memo_type_to_code, ID16_OFFSET, ISSUER_TAG_OFFSET, ISSUER_TAG_SIZE, MEMO_BYTES, SALT_OFFSET,
    SALT_SIZE,
};
use super::ulid::ulid_to_bytes16;
use crate::types::MemoType;
use crate::MemoError;

/// Parameters for encoding a v1 memo.
#[derive(Debug)]
pub struct EncodeMemoV1Params {
    pub memo_type: MemoType,
    /// Issuer namespace tag (u64). Derive with [`crate::issuer_tag_from_namespace`].
    pub issuer_tag: u64,
    /// 26-character Crockford base32 ULID string.
    pub ulid: String,
    /// Optional 7-byte salt. `None` encodes as seven zero bytes.
    pub salt: Option<[u8; 7]>,
}

/// Encode memo fields into a 32-byte hex string per TEMPO-RECONCILE-MEMO-001.
///
/// Layout: `[type:1][issuerTag:8][id16:16][salt:7]` = 32 bytes.
///
/// Returns `"0x"` + 64 lowercase hex characters.
///
/// # Errors
/// Returns [`MemoError`] if the ULID is invalid (wrong length, bad chars).
///
/// # Example
/// ```
/// use tempo_reconcile::{encode_memo_v1, issuer_tag_from_namespace, EncodeMemoV1Params};
/// use tempo_reconcile::MemoType;
///
/// let tag = issuer_tag_from_namespace("my-app");
/// let memo = encode_memo_v1(&EncodeMemoV1Params {
///     memo_type: MemoType::Invoice,
///     issuer_tag: tag,
///     ulid: "01MASW9NF6YW40J40H289H858P".to_string(),
///     salt: None,
/// }).unwrap();
/// assert!(memo.starts_with("0x"));
/// assert_eq!(memo.len(), 66);
/// ```
pub fn encode_memo_v1(params: &EncodeMemoV1Params) -> Result<String, MemoError> {
    let id16 = ulid_to_bytes16(&params.ulid)?;

    let salt = params.salt.unwrap_or([0u8; SALT_SIZE]);

    let mut buf = [0u8; MEMO_BYTES];

    buf[0] = memo_type_to_code(&params.memo_type);

    buf[ISSUER_TAG_OFFSET..ISSUER_TAG_OFFSET + ISSUER_TAG_SIZE]
        .copy_from_slice(&params.issuer_tag.to_be_bytes());

    buf[ID16_OFFSET..ID16_OFFSET + 16].copy_from_slice(&id16);
    buf[SALT_OFFSET..SALT_OFFSET + SALT_SIZE].copy_from_slice(&salt);

    Ok(format!("0x{}", hex::encode(buf)))
}

/// Generate 7 cryptographically random bytes for use as a memo salt.
#[cfg(feature = "rand")]
pub fn random_salt() -> [u8; 7] {
    use rand::RngCore;
    let mut buf = [0u8; 7];
    rand::thread_rng().fill_bytes(&mut buf);
    buf
}
